notebooks/08B_modeling_parallel_sorting_data.ipynb

```diff
@@ -0,0 +1,561 @@
+    "    )  # Recall\n",
+    "    f1 = metrics.f1_score(\n",
+    "        y_true=y_true, y_pred=y_pred, average=\"weighted\", zero_division=0\n",
+    "    )  # F1\n",
+    "    conf_matrix = metrics.confusion_matrix(\n",
+    "        y_true=y_true, y_pred=y_pred\n",
+    "    )  # Confusion matrix\n",
+    "    unravel_true, unravel_pred = unravel_true_pred(y_true, y_pred)\n",
+    "    y_proba = y_proba.ravel()\n",
+    "    try:  # case where class sample didnt get into data\n",
+    "        roc_auc_score = metrics.roc_auc_score(\n",
+    "            y_true=unravel_true,\n",
+    "            y_score=y_proba,\n",
+    "            multi_class=\"ovr\",\n",
+    "            average=\"weighted\",\n",
+    "        )\n",
+    "    except ValueError:\n",
+    "        roc_auc_score = 0\n",
+    "    return acc, pre, rec, f1, conf_matrix, roc_auc_score"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Modeling Preparation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3A. Unpickle the Thicket from part A\n",
+    "\n",
+    "Running `08A_composing_parallel_sorting_data.ipynb` is necessary before this step."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tk = th.Thicket.from_pickle(\"thicket-modeldata.pkl\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3B. Concatenate Features\n",
+    "To match the expected format of the Scikit-learn models, we concatenate the features together such that each row in the `model_data` DataFrame is a data sample (profile). We can achieve this desired format by using `pd.DataFrame.unstack()` to pivot the node index into the column labels, and using `pd.concat()` to concatenate the results. \n",
+    "\n",
+    "As a simple example: applying the unstack operation to a MultiIndex DataFrame (node and profile) with two unique values for node and two columns will result in a DataFrame with one index level (profile) and 4 columns (2 nodes x 2 columns). We are essentially extending the DataFrame on the column axis."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_data = pd.concat(\n",
+    "    [\n",
+    "        tk.dataframe.loc[tk.perf_idx].unstack(level=\"node\"),\n",
+    "        tk.dataframe.loc[tk.presence_idx].unstack(level=\"node\"),\n",
+    "        tk.metadata[\"Algorithm\"]\n",
+    "    ],\n",
+    "    axis=\"columns\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3C. Define Categorical and Numerical Columns\n",
+    "\n",
+    "We manually define the two categorical columns and then define the numerical columns, which are by definition the remaining columns in the data."
```

Comment:
Maybe here you can just remind them that "Node presence" was a derived feature from the first notebook in this description. I'm still not quite understanding what the feature means, but that is probably described in your paper?

Reply:
2248c132 - rename feature from node presence to "Present" to match paper

73c70cd6 - update 3c text

Yes it's described in the paper as "Present"